### PR TITLE
KT-30970 No warning for empty `if` operator and `also`method

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-30970

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import com.intellij.psi.PsiComment
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.idea.inspections.collections.isCalling
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.allChildren
+
+class ControlFlowWithEmptyBodyInspection : AbstractKotlinInspection() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean, session: LocalInspectionToolSession) = object : KtVisitorVoid() {
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: `buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean)`

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            val elseKeyword = expression.elseKeyword
+            if (elseKeyword != null && expression.`else`.isEmptyBody()) {
+                holder.registerProblem(elseKeyword)
+            }
+        }
+
+        override fun visitWhenExpression(expression: KtWhenExpression) {
+            super.visitWhenExpression(expression)
+            if (expression.entries.isNotEmpty() || expression.hasComment()) return
+            val keyword = expression.allChildren.firstOrNull { it.node.elementType == KtTokens.WHEN_KEYWORD } ?: return
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: Why not `expression.whenKeyword`?

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import org.jetbrains.kotlin.idea.inspections.collections.isCalling
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.allChildren
+
+class ControlFlowWithEmptyBodyInspection : AbstractKotlinInspection() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean, session: LocalInspectionToolSession) = object : KtVisitorVoid() {
+        override fun visitIfExpression(expression: KtIfExpression) {
+            super.visitIfExpression(expression)
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: Redundant call.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                holder.registerProblem(expression.ifKeyword)
+            }
+            val elseKeyword = expression.elseKeyword
+            if (elseKeyword != null && expression.`else`.isEmptyBody()) {
+                holder.registerProblem(elseKeyword)
+            }
+        }
+
+        override fun visitWhenExpression(expression: KtWhenExpression) {
+            super.visitWhenExpression(expression)
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: Same as above.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            if (expression.entries.isNotEmpty() || expression.hasComment()) return
+            val keyword = expression.allChildren.firstOrNull { it.node.elementType == KtTokens.WHEN_KEYWORD } ?: return
+            holder.registerProblem(keyword)
+        }
+
+        override fun visitForExpression(expression: KtForExpression) {
+            super.visitForExpression(expression)
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: Same as above.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            if (expression.body.isEmptyBody()) {
+                holder.registerProblem(expression.forKeyword)
+            }
+        }
+
+        override fun visitWhileExpression(expression: KtWhileExpression) {
+            super.visitWhileExpression(expression)
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: Same as above.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            val keyword = expression.allChildren.firstOrNull { it.node.elementType == KtTokens.WHILE_KEYWORD } ?: return
+            if (expression.body.isEmptyBody()) {
+                holder.registerProblem(keyword)
+            }
+        }
+
+        override fun visitDoWhileExpression(expression: KtDoWhileExpression) {
+            super.visitDoWhileExpression(expression)
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: Same as above.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            val keyword = expression.allChildren.firstOrNull { it.node.elementType == KtTokens.DO_KEYWORD } ?: return
+            if (expression.body.isEmptyBody()) {
+                holder.registerProblem(keyword)
+            }
+        }
+
+        override fun visitCallExpression(expression: KtCallExpression) {
+            super.visitCallExpression(expression)
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: Same as above.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        if (this is KtBlockExpression && statements.isEmpty() && !hasComment()) return true
+        return false
+    }
+
+    private fun KtExpression.hasComment(): Boolean {
+        return this.allChildren.any { it is PsiComment }
+    }
+
+    private fun ProblemsHolder.registerProblem(keyword: PsiElement) {
+        registerProblem(keyword, "'${keyword.text}' has empty body")
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: We should suggest a fix to remove an empty construction.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+import org.jetbrains.kotlin.psi.psiUtil.allChildren
+import org.jetbrains.kotlin.psi.psiUtil.getQualifiedExpressionForSelector
+import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
+import org.jetbrains.kotlin.psi.psiUtil.startOffset
+
+class ControlFlowWithEmptyBodyInspection : AbstractKotlinInspection() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) = object : KtVisitorVoid() {
+        override fun visitIfExpression(expression: KtIfExpression) {
+            if (expression.then.isEmptyBody()) {
+                val fix = if (expression.hasEmptyBranches()) RemoveElementFix() else null
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: We can't remove an expression if the condition has a side effect. We should replace expression with the condition. 
For example, Java suggest `Extract side effect` in this case.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            }
+            val elseKeyword = expression.elseKeyword
+            if (elseKeyword != null && expression.`else`.isEmptyBody()) {
+                holder.registerProblem(expression, elseKeyword, RemoveElementFix(isIfElseBranch = true))
+            }
+        }
+
+        override fun visitWhenExpression(expression: KtWhenExpression) {
+            if (expression.entries.isNotEmpty() || expression.hasComment()) return
+            holder.registerProblem(expression, expression.whenKeyword)
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: Same as above.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        }
+
+        override fun visitForExpression(expression: KtForExpression) {
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt]: We shoudn't suggest fix for `for`, `while` and `do-while`.

**Comment by dimonchik0036** [comment]
Creating the correct quickfix is harder than I thought. There are many different cases that need to be covered. Do you want to cover all cases from Java? If not, i can merge only inspection without quickfix.

**Comment by t-kameyama** [comment]
Please merge only inspection.

**Comment by dimonchik0036** [comment]
Merged, thanks! 65f06454be3661333bd452881f2f6cb80e756b02

**Pull request closed without merging**